edb/pgsql/resolver/command.py

```diff
@@ -339,6 +369,409 @@ def _preprocess_insert_stmt(
+
+    # prepare anchors for inserted value columns
+    value_name = ctx.alias_generator.get('ins_val')
+    iterator_name = ctx.alias_generator.get('ins_iter')
+    source_id = irast.PathId.from_type(
+        ctx.schema,
+        sub_source,
+        typename=sn.QualName('__derived__', value_name),
+        env=None,
+    )
+    link_ref = irtypeutils.ptrref_from_ptrcls(
+        schema=ctx.schema, ptrcls=link, cache=None, typeref_cache=None
+    )
+    value_id: irast.PathId = source_id.extend(ptrref=link_ref)
+
+    value_ql: qlast.PathElement = (
+        qlast.IRAnchor(name=value_name)
+        if is_value_single
+        else qlast.ObjectRef(name=iterator_name)
+    )
+
+    # a phantom relation that is supposed to hold the inserted value
+    # (in the resolver, this will be replaced by the real value relation)
+    value_cte_name = ctx.alias_generator.get('ins_value')
+    value_rel = pgast.Relation(name=value_cte_name)
+    value_columns: List[Tuple[str, bool]] = []
+    for expected_col in expected_columns:
+        if expected_col.name == 'source':
+            ptr_name = 'source'
+            is_link = True
+
+            output_var = pgast.ColumnRef(name=(ptr_name,), nullable=True)
+            value_rel.path_outputs[(source_id, pgce.PathAspect.VALUE)] = (
+                output_var
+            )
+            value_rel.path_outputs[(source_id, pgce.PathAspect.IDENTITY)] = (
+                output_var
+            )
+        elif expected_col.name == 'target':
+            ptr_name = 'target'
+            is_link = True
+
+            ptr_id = value_id.tgt_path()
+            output_var = pgast.ColumnRef(name=(ptr_name,), nullable=True)
+            value_rel.path_outputs[(ptr_id, pgce.PathAspect.VALUE)] = output_var
+            value_rel.path_outputs[(ptr_id, pgce.PathAspect.IDENTITY)] = (
+                output_var
+            )
+        else:
+            # link pointer
+            ptr_name = expected_col.name
+            ptr = link.maybe_get_ptr(ctx.schema, sn.UnqualName(ptr_name))
+            assert ptr
+            is_link = False
+
+            ptr_id = _get_ptr_id(value_id.ptr_path(), ptr, ctx)
+            output_var = pgast.ColumnRef(name=(ptr_name,), nullable=True)
+            value_rel.path_outputs[(ptr_id, pgce.PathAspect.VALUE)] = output_var
+
+        value_columns.append((ptr_name, is_link))
+
+    # source needs an iterator column, so we need to invent one
+    # Here we only decide on the name of that iterator column, the actual column
+    # is generated later, when resolving the DML stmt.
+    value_iterator = ctx.alias_generator.get('iter')
+    output_var = pgast.ColumnRef(name=(value_iterator,))
+    value_rel.path_outputs[(source_id, pgce.PathAspect.ITERATOR)] = output_var
+    value_rel.path_outputs[(value_id, pgce.PathAspect.ITERATOR)] = output_var
+
+    # construct the EdgeQL DML AST
+    sub_name = link.get_name(ctx.schema)
+    assert isinstance(sub_name, sn.QualName)
```

Comment:
These asserts shouldn't be needed, I think? `Link` is a `QualifiedObject`, so its name is always a QualName.
(If it is more about convincing `pylance` than `mypy`, maybe put an annotation on the variable instead of adding an assert?)

Reply:
You were right.

I think I've manually went and looked at the definition of `name = so.Field(...)` on `so.Object` and derived the assert from there.